File: 프로그래머스/연습문제/Lv1/햄버거 만들기/햄버거 만들기.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(vector<int> ingredient) {
    stack<int> st;
    int cnt = 0;
    
    for(int i=0; i<ingredient.size(); i++){
        if(st.size() < 3) st.push(ingredient[i]);
        else if(ingredient[i] == 1){
            if(st.top() != 3) st.push(ingredient[i]);
            else{
                st.pop();
                if(st.top() != 2){
                    st.push(3);
                    st.push(ingredient[i]);
                }
                else{
                    st.pop();
                    if(st.top() != 1){
                        st.push(2); st.push(3);
                        st.push(ingredient[i]);
                    }
                    else{
                        st.pop();
                        cnt++;
                    }
                }
            }
        }
        else st.push(ingredient[i]);
    }
    
    return cnt;
}

File: 프로그래머스/연습문제/Lv1/햄버거 만들기/햄버거 만들기.py
def solution(ingredient):
    st = []
    cnt = 0

    for ing in ingredient:
        if len(st)<3 or ing!=1: st.append(ing)
        elif st[-1]==3 and st[-2]==2 and st[-3]==1:
            st.pop(-1); st.pop(-1); st.pop(-1)
            cnt += 1
        else: st.append(ing)

    return cnt
